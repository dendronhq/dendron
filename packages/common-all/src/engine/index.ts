export * from "./EngineEventEmitter";
export * from "./EngineV3Base";
