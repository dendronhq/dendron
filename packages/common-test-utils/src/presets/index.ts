export * from "./plugin-core";
export * from "./utils";
export * from "./notes";
export * from "./schemas";
