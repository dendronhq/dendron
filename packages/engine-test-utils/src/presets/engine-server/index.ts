import { ENGINE_CONFIG_PRESETS } from "./config";
import { ENGINE_DELETE_PRESETS } from "./delete";
import { ENGINE_INFO_PRESETS } from "./info";
import { ENGINE_INIT_PRESETS } from "./init";
import { ENGINE_GET_NOTE_BLOCKS_PRESETS } from "./getNoteBlocks";
import NOTE_REF from "./note-refs";
import { ENGINE_QUERY_PRESETS } from "./query";
import { ENGINE_RENAME_PRESETS } from "./rename";
import { ENGINE_BULK_WRITE_NOTES_PRESETS } from "./bulkWriteNotes";
import { ENGINE_RENDER_PRESETS } from "./render";
import { ENGINE_WRITE_PRESETS, ENGINE_WRITE_PRESETS_MULTI } from "./write";
import _ from "lodash";
import { TestPresetEntryV4 } from "@dendronhq/common-test-utils";

export { ENGINE_HOOKS, ENGINE_HOOKS_BASE, ENGINE_HOOKS_MULTI } from "./utils";
export { ENGINE_RENAME_PRESETS };
export { ENGINE_QUERY_PRESETS };
export { ENGINE_WRITE_PRESETS };
export { ENGINE_CONFIG_PRESETS };

export const ENGINE_SERVER = {
  NOTE_REF,
  ENGINE_WRITE_PRESETS,
  ENGINE_INIT_PRESETS,
  ENGINE_DELETE_PRESETS,
  ENGINE_INFO_PRESETS,
  ENGINE_RENAME_PRESETS,
  ENGINE_GET_NOTE_BLOCKS_PRESETS,
  ENGINE_QUERY_PRESETS,
  ENGINE_BULK_WRITE_NOTES_PRESETS,
  ENGINE_RENDER_PRESETS,
};

type TestPresetEntry = TestPresetEntryV4;
type TestPresetDict = { [key: string]: TestPresetEntry };

/**
 * 
 @example
 *  test("", async () => {
 *    const TestCase= getPreset({presets: ENGINE_PRESETS, nodeType, presetName: "init", key: 'BAD_SCHEMA'})
 *    const { testFunc, ...opts } = TestCase;;
 *    await runEngineTestV5(testFunc, { ...opts, expect });
 *});
 * @param param0 
 * @returns 
 */
export const getPreset = ({
  presets,
  presetName,
  nodeType,
  key,
}: {
  presets: typeof ENGINE_PRESETS;
  presetName: string;
  nodeType: "SCHEMAS" | "NOTES";
  key: string;
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  const out = _.get(ent.presets[nodeType], key);
  if (!out) {
    throw Error(`no key ${key} found in ${presetName}`);
  }
  return out;
};

export const getPresetMulti = ({
  presets,
  presetName,
  nodeType,
  key,
}: {
  presets: typeof ENGINE_PRESETS_MULTI;
  presetName: string;
  nodeType: "NOTES";
  key: string;
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  const out = _.get(ent.presets[nodeType], key);
  if (!out) {
    throw Error(`no key ${key} found in ${presetName}`);
  }
  return out;
};

export const getPresetGroup = ({
  presets,
  presetName,
  nodeType,
}: {
  presets: typeof ENGINE_PRESETS;
  presetName: string;
  nodeType: "SCHEMAS" | "NOTES";
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  return ent.presets[nodeType] as TestPresetDict;
};

// ^iygzn9r2758w
export const ENGINE_PRESETS = [
  { name: "write", presets: ENGINE_SERVER.ENGINE_WRITE_PRESETS },
  {
    name: "bulkWriteNotes",
    presets: ENGINE_SERVER.ENGINE_BULK_WRITE_NOTES_PRESETS,
  },
  { name: "delete", presets: ENGINE_SERVER.ENGINE_DELETE_PRESETS },
  { name: "rename", presets: ENGINE_SERVER.ENGINE_RENAME_PRESETS },
  { name: "init", presets: ENGINE_SERVER.ENGINE_INIT_PRESETS },
  { name: "query", presets: ENGINE_SERVER.ENGINE_QUERY_PRESETS },
  { name: "info", presets: ENGINE_SERVER.ENGINE_INFO_PRESETS },
  {
    name: "getNoteBlocks",
    presets: ENGINE_SERVER.ENGINE_GET_NOTE_BLOCKS_PRESETS,
  },
  { name: "render", presets: ENGINE_SERVER.ENGINE_RENDER_PRESETS },
];

export const ENGINE_PRESETS_MULTI = [
  { name: "write", presets: ENGINE_WRITE_PRESETS_MULTI },
  { name: "delete", presets: ENGINE_SERVER.ENGINE_DELETE_PRESETS },
  { name: "rename", presets: ENGINE_SERVER.ENGINE_RENAME_PRESETS },
  { name: "init", presets: ENGINE_SERVER.ENGINE_INIT_PRESETS },
  { name: "query", presets: ENGINE_SERVER.ENGINE_QUERY_PRESETS },
];
