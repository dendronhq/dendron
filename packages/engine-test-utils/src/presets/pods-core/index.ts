import JSON from "./json";
import MARKDOWN from "./markdown";

export const PODS_CORE = {
  JSON,
  MARKDOWN,
};

export const PODS_PRESETS = [{ name: "json", presets: JSON }];
