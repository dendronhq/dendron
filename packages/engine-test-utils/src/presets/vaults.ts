import { DVault } from "@dendronhq/common-all";

export const VAULTS = {
  MULTI_VAULT_WITH_THREE_VAULTS: () => {
    return [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      { fsPath: "vault3", name: "vaultThree" },
    ] as DVault[];
  },
};
