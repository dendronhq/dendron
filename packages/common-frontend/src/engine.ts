import {
  IntermediateDendronConfig,
  DEngineClient,
  DVault,
  NotePropsDict,
  SchemaModuleDict,
} from "@dendronhq/common-all";
import { EngineState } from "./features/engine/slice";

// @ts-ignore
export class WebEngine implements DEngineClient {
  public notes: NotePropsDict;
  public wsRoot: string;
  public schemas: SchemaModuleDict;
  public configRoot: string;
  public vaults: DVault[];
  // public links: DLink[];
  // public fuseEngine: FuseEngine;
  // public api: DendronAPI;
  // public history?: HistoryService;
  // public logger: DLogger;
  public config: IntermediateDendronConfig;
  // public hooks: DHookDict;

  static create(state: Required<EngineState>) {
    return new WebEngine(state);
  }
  constructor(state: Required<EngineState>) {
    const { notes, wsRoot, schemas, vaults, config } = state;
    this.notes = notes;
    this.wsRoot = wsRoot;
    this.schemas = schemas;
    this.vaults = vaults;
    this.configRoot = wsRoot;
    this.config = config;
  }
}
