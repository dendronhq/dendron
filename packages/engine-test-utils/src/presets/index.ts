import { DVault, VaultUtils } from "@dendronhq/common-all";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import { WorkspaceOpts } from "../engine";

export * from "./engine-server";
export * from "./pods-core";
export * from "./vaults";

export enum SETUP_HOOK_KEYS {
  /**
   * alpha: link(beta)
   * beta: link(alpha)
   */
  WITH_LINKS = "WITH_LINKS",
}

const createLink = (fname: string, opts: { vaultPrefix?: DVault }) => {
  let cVaultPrefix = "";
  if (opts.vaultPrefix) {
    cVaultPrefix = VaultUtils.toURIPrefix(opts.vaultPrefix) + "/";
  }
  return `[[${cVaultPrefix}${fname}]]`;
};

export async function callSetupHook(
  key: SETUP_HOOK_KEYS,
  opts: {
    workspaceType: "single" | "multi";
    withVaultPrefix?: boolean;
  } & WorkspaceOpts
) {
  const { workspaceType, vaults, wsRoot, withVaultPrefix } = opts;
  const isMultiVault = workspaceType !== "single";
  const cVaults = isMultiVault ? vaults : [vaults[0], vaults[0]];

  // WITH LINKS
  if (key === SETUP_HOOK_KEYS.WITH_LINKS) {
    const link1 = createLink("beta", {
      vaultPrefix: withVaultPrefix ? cVaults[1] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "alpha",
      wsRoot,
      vault: cVaults[0],
      body: link1,
    });
    const link2 = createLink("alpha", {
      vaultPrefix: withVaultPrefix ? cVaults[0] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "beta",
      wsRoot,
      vault: cVaults[1],
      body: link2,
    });
    if (isMultiVault) {
      await NoteTestUtilsV4.createNote({
        fname: "beta",
        wsRoot,
        vault: cVaults[0],
        body: link2,
        genRandomId: true,
      });
    }
  } else {
    throw Error("not supported key");
  }
}
