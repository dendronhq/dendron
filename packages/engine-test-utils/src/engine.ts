import {
  CleanDendronSiteConfig,
  CONSTANTS, DendronConfig, DEngineClient,
  DVault,
  DWorkspace,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings
} from "@dendronhq/common-all";
import {
  getDurationMilliseconds,
  tmpDir,
  vault2Path
} from "@dendronhq/common-server";
import {
  GenTestResults,
  PostSetupHookFunction,
  PreSetupHookFunction,
  RunEngineTestFunctionOpts,
  RunEngineTestFunctionV4,
  runJestHarnessV2,
  SetupHookFunction,
  SetupTestFunctionV4,
  sinon,
  TestResult
} from "@dendronhq/common-test-utils";
import { LaunchEngineServerCommand } from "@dendronhq/dendron-cli";
import {
  createEngine as engineServerCreateEngine,
  DConfig,
  WorkspaceService
} from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import os from "os";
import path from "path";
import { SinonStub } from "sinon";
import { ENGINE_HOOKS } from "./presets";
import { GitTestUtils } from "./utils";

export type TestSetupWorkspaceOpts = {
  /**
   * Vaults to initialize engine with
   * Defaults to following if not set
   * [
   *    { fsPath: "vault1" },
   *    { fsPath: "vault2" },
   *    { fsPath: "vault3", name: "vaultThree" },
   *  ]
   */
  vaults?: DVault[];
  /**
   * Modify dendron config before initialization
   */
  modConfigCb?: (config: DendronConfig) => DendronConfig;
};

export type AsyncCreateEngineFunction = (
  opts: WorkspaceOpts
) => Promise<{engine: DEngineClient, port?: number}>;

/**
 * Create an {@link DendronEngine}
 */
export async function createEngineFromEngine(opts: WorkspaceOpts) {
  return {engine: engineServerCreateEngine(opts) as DEngineClient, port: undefined};
}

export { DEngineClient, DVault, WorkspaceOpts };

/**
 * Create a server
 * @param opts
 * @returns
 */
// @ts-ignore
export async function createServer(opts: WorkspaceOpts) {
  return await new LaunchEngineServerCommand().enrichArgs({
    wsRoot: opts.wsRoot,
  });
}

/**
 * Create an {@link DendronEngineClient}
 */
export async function createEngineFromServer(
  opts: WorkspaceOpts
): Promise<any> {
  const { engine, port } = await createServer(opts);
  await engine.init();
  return {engine, port};
}

export function createSiteConfig(
  opts: Partial<CleanDendronSiteConfig> &
    Required<Pick<CleanDendronSiteConfig, "siteRootDir" | "siteHierarchies">>
): CleanDendronSiteConfig {
  let copts = {
    siteNotesDir: "docs",
    siteUrl: "https://localhost:8080",
    ...opts,
  };
  return {
    ...copts,
    siteIndex: DConfig.getSiteIndex(copts),
  };
}

/**
 *
 * @param opts.asRemote: add git repo
 * @param opts.wsRoot: override given wsRoot
 * @returns
 */
export async function setupWS(opts: {
  vaults: DVault[];
  workspaces?: DWorkspace[];
  asRemote?: boolean;
  wsRoot?: string;
}) {
  const wsRoot = opts.wsRoot || tmpDir().name;
  const ws = new WorkspaceService({ wsRoot });
  ws.createConfig();
  const config = ws.config;
  let vaults = await Promise.all(
    opts.vaults.map(async (vault) => {
      await ws.createVault({ vault, config, updateConfig: false });
      return vault;
    })
  );
  config.vaults = _.sortBy(config.vaults, "fsPath");
  if (config.site.duplicateNoteBehavior) {
    config.site.duplicateNoteBehavior.payload = (
      config.site.duplicateNoteBehavior.payload as string[]
    ).sort();
  }
  ws.setConfig(config);
  if (opts.workspaces) {
    const vaultsFromWs = await _.reduce(
      opts.workspaces,
      async (resp, ent) => {
        await resp;
        await WorkspaceService.createWorkspace({
          wsRoot: path.join(wsRoot, ent.name),
          vaults: ent.vaults,
        });
        return ws.addWorkspace({ workspace: ent });
      },
      Promise.resolve({ vaults: [] } as { vaults: DVault[] })
    );
    vaults = vaults.concat(vaultsFromWs.vaults);
  }
  if (opts.asRemote) {
    await GitTestUtils.createRepoWithReadme(wsRoot);
  }
  return { wsRoot, vaults };
}

export type RunEngineTestV5Opts = {
  preSetupHook?: SetupHookFunction;
  createEngine?: AsyncCreateEngineFunction;
  extra?: any;
  expect: any;
  workspaces?: DWorkspace[];
  setupOnly?: boolean;
  initGit?: boolean;
  initHooks?: boolean;
  addVSWorkspace?: boolean;
  /**
   * Path to preset wsRoot
   */
  wsRoot?: string;
} & TestSetupWorkspaceOpts;

export type RunEngineTestFunctionV5<T = any> = (
  opts: RunEngineTestFunctionOpts & { extra?: any; engineInitDuration: number, port?: number }
) => Promise<TestResult[] | void | T>;

export class TestPresetEntryV5 {
  public preSetupHook: PreSetupHookFunction;
  public postSetupHook: PostSetupHookFunction;
  public testFunc: RunEngineTestFunctionV4;
  public extraOpts: any;
  public setupTest?: SetupTestFunctionV4;
  public genTestResults?: GenTestResults;
  public vaults: DVault[];
  public workspaces: DWorkspace[];

  constructor(
    func: RunEngineTestFunctionV5,
    opts?: {
      preSetupHook?: PreSetupHookFunction;
      postSetupHook?: PostSetupHookFunction;
      extraOpts?: any;
      setupTest?: SetupTestFunctionV4;
      genTestResults?: GenTestResults;
      vaults?: DVault[];
    }
  ) {
    let {
      preSetupHook,
      postSetupHook,
      extraOpts,
      setupTest,
      genTestResults,
      workspaces,
    } = _.defaults(opts, {
      workspaces: [],
    });
    this.preSetupHook = preSetupHook ? preSetupHook : async () => {};
    this.postSetupHook = postSetupHook ? postSetupHook : async () => {};
    this.testFunc = _.bind(func, this);
    this.extraOpts = extraOpts;
    this.setupTest = setupTest;
    this.genTestResults = _.bind(
      genTestResults ? genTestResults : async () => [],
      this
    );
    this.workspaces = workspaces;
    this.vaults = opts?.vaults || [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      {
        name: "vaultThree",
        fsPath: "vault3",
      },
    ];
  }
}

/**
 *
 * To create empty workspace, initilizae with `vaults = []`
 * @param func
 * @param opts.vaults: By default, initiate 3 vaults {vault1, vault2, (vault3, "vaultThree")}
 * @param opts.preSetupHook: By default, initiate empty
 * @param opts.wsRoot: Override the randomly generated test directory for the wsRoot
 * @returns
 */
export async function runEngineTestV5(
  func: RunEngineTestFunctionV5,
  opts: RunEngineTestV5Opts
): Promise<any> {
  const {
    preSetupHook,
    extra,
    vaults: vaultsInit,
    createEngine,
    initGit,
    workspaces,
    addVSWorkspace,
  } = _.defaults(opts, {
    preSetupHook: async ({}) => {},
    postSetupHook: async ({}) => {},
    createEngine: createEngineFromEngine,
    extra: {},
    // third vault has diff name
    vaults: [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      { fsPath: "vault3", name: "vaultThree" },
    ],
    addVSWorkspace: false,
  });
  try {
    // make sure tests don't overwrite local homedir contents
    TestEngineUtils.mockHomeDir();
    const { wsRoot, vaults } = await setupWS({
      vaults: vaultsInit,
      workspaces,
      wsRoot: opts.wsRoot,
    });
    if ((opts.initHooks, vaults)) {
      fs.ensureDirSync(path.join(wsRoot, CONSTANTS.DENDRON_HOOKS_BASE));
    }
    await preSetupHook({ wsRoot, vaults });
    const resp = await createEngine({ wsRoot, vaults });
    const engine = resp.engine;
    const start = process.hrtime();
    const initResp = await engine.init();
    if (addVSWorkspace) {
      fs.writeJSONSync(
        path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME),
        {
          folders: vaults.map((ent) => ({
            path: ent.fsPath,
            name: ent.name,
          })) as WorkspaceFolderRaw[],
          settings: {},
          extensions: {},
        } as WorkspaceSettings,
        { spaces: 4 }
      );
    }
    const engineInitDuration = getDurationMilliseconds(start);
    const testOpts = {
      wsRoot,
      vaults,
      engine,
      initResp,
      port: resp.port,
      extra,
      config: engine,
      engineInitDuration,
    };
    if (initGit) {
      await GitTestUtils.createRepoForWorkspace(wsRoot);
      await Promise.all(
        vaults.map((vault) => {
          return GitTestUtils.createRepoWithReadme(
            vault2Path({ vault, wsRoot })
          );
        })
      );
    }
    if (opts.setupOnly) {
      return testOpts;
    }
    const results = (await func(testOpts)) || [];
    await runJestHarnessV2(results, expect);
    return { opts: testOpts, resp: undefined, wsRoot };
  } finally {
    // restore sinon so other tests can keep running
    sinon.restore();
  }
}

export function testWithEngine(
  prompt: string,
  func: RunEngineTestFunctionV4,
  opts?: Omit<RunEngineTestV5Opts, "expect"> & { only?: boolean }
) {
  if (opts?.only) {
    return test.only(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  } else {
    return test(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  }
}

export class TestEngineUtils {
  static mockHomeDir(dir?: string): SinonStub {
    if (_.isUndefined(dir)) dir = tmpDir().name;
    return sinon.stub(os, "homedir").returns(dir);
  }

  static vault1(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault1" })!;
  }

  static vault2(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault2" })!;
  }

  static vault3(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault3" })!;
  }
}
