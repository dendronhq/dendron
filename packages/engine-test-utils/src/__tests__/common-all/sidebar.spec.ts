import {
  getSidebar,
  DefaultSidebar,
  NoteDictsUtils,
} from "@dendronhq/common-all";
import { runEngineTestV5 } from "../../engine";
import { ENGINE_HOOKS } from "../../presets";

/*
 * See [[dendron://dendron.dendron-site/dendron.topic.publish.sidebar#complex-sidebar-example]] for sidebar docs and example
 */

describe("GIVEN sidebar config input", () => {
  describe("WHEN providing empty config", () => {
    test("THEN return empty sidebar object", async () => {
      await runEngineTestV5(
        async ({ engine }) => {
          const engineNotes = await engine.findNotes({ excludeStub: false });
          const sidebarResp = getSidebar([], {
            notes: NoteDictsUtils.createNotePropsByIdDict(engineNotes),
          });
          expect(sidebarResp._unsafeUnwrap()).toHaveLength(0);
        },
        {
          expect,
          preSetupHook: ENGINE_HOOKS.setupBasic,
        }
      );
    });
  });

  describe("WHEN providing DefaultSidebar", () => {
    test("THEN return sidebar object", async () => {
      await runEngineTestV5(
        async ({ engine }) => {
          const engineNotes = await engine.findNotes({ excludeStub: false });
          const sidebarResp = getSidebar(DefaultSidebar, {
            notes: NoteDictsUtils.createNotePropsByIdDict(engineNotes),
          });
          expect(sidebarResp._unsafeUnwrap()).toMatchSnapshot();
          // Expect DefaultSidebar to resolve into 2 sidebarItem entries
          expect(sidebarResp._unsafeUnwrap()).toHaveLength(2);
          // Expect `foo.md` to have the child `foo.ch1.md`
          expect(sidebarResp._unsafeUnwrap()).toHaveProperty(
            "[1].items[0].id",
            "foo.ch1"
          );
        },
        {
          expect,
          preSetupHook: ENGINE_HOOKS.setupBasic,
        }
      );
    });
  });

  describe("WHEN providing autogenerated sidebar item", () => {
    test("THEN return sidebar slice", async () => {
      await runEngineTestV5(
        async ({ engine }) => {
          const engineNotes = await engine.findNotes({ excludeStub: false });
          const sidebarResp = getSidebar(
            [{ type: "autogenerated", id: "foo" }],
            {
              notes: NoteDictsUtils.createNotePropsByIdDict(engineNotes),
            }
          );
          expect(sidebarResp._unsafeUnwrap()).toMatchSnapshot();
          // Expect to resolve into sidebar slice with two children
          expect(sidebarResp._unsafeUnwrap()).toHaveLength(2);
          // Expect to contain deep nested note
          expect(sidebarResp._unsafeUnwrap()).toHaveProperty(
            "[0].items[0].items[0].id",
            "foo.ch1.gch1.ggch1"
          );
        },
        {
          expect,
          preSetupHook: ENGINE_HOOKS.setupHierarchyForLookupTests,
        }
      );
    });
  });

  describe("WHEN providing custom sidebar", () => {
    test("THEN return sidebar object", async () => {
      await runEngineTestV5(
        async ({ engine }) => {
          const engineNotes = await engine.findNotes({ excludeStub: false });
          const sidebarResp = getSidebar(
            [
              {
                type: "note",
                label: "Foo label",
                id: "foo",
              },
              {
                type: "category",
                label: "Some category label",
                link: { type: "note", id: "bar" },
                items: [
                  {
                    type: "autogenerated",
                    id: "foo.ch1",
                  },
                  {
                    type: "note",
                    label: "Some deep note",
                    id: "goo.ends-with-ch1.no-ch1-by-itself",
                  },
                ],
              },
            ],
            { notes: NoteDictsUtils.createNotePropsByIdDict(engineNotes) }
          );
          expect(sidebarResp._unsafeUnwrap()).toMatchSnapshot();
          expect(sidebarResp._unsafeUnwrap()).toHaveProperty(
            "[1].items[0].items[0].id",
            "foo.ch1.gch1.ggch1"
          );
          expect(sidebarResp._unsafeUnwrap()).toHaveProperty(
            "[1].items[2].id",
            "goo.ends-with-ch1.no-ch1-by-itself"
          );
        },
        {
          expect,
          preSetupHook: ENGINE_HOOKS.setupHierarchyForLookupTests,
        }
      );
    });
  });

  describe("WHEN nav_exclude_children is enabled", () => {
    test("THEN return sidebar object", async () => {
      await runEngineTestV5(
        async ({ engine }) => {
          const domain = (await engine.getNote("foo")).data!;
          domain.custom.nav_exclude_children = true;
          await engine.writeNote(domain, { metaOnly: true });
          const engineNotes = await engine.findNotes({ excludeStub: false });
          const sidebarResp = getSidebar(
            [{ type: "autogenerated", id: "root" }],
            {
              notes: NoteDictsUtils.createNotePropsByIdDict(engineNotes),
            }
          );
          expect(sidebarResp._unsafeUnwrap()).toMatchSnapshot();
          // Expect foo note to have no children
          const fooSidebarItem = sidebarResp._unsafeUnwrap()?.[1];
          expect(fooSidebarItem).toHaveProperty("items", []);
        },
        {
          expect,
          preSetupHook: ENGINE_HOOKS.setupBasic,
        }
      );
    });
  });

  describe("WHEN has_collection is enabled", () => {
    test("THEN return sidebar object", async () => {
      await runEngineTestV5(
        async ({ engine }) => {
          const domain = (await engine.getNote("foo")).data!;
          domain.custom.has_collection = true;
          await engine.writeNote(domain, { metaOnly: true });
          const engineNotes = await engine.findNotes({ excludeStub: false });
          const sidebarResp = getSidebar(
            [{ type: "autogenerated", id: "root" }],
            {
              notes: NoteDictsUtils.createNotePropsByIdDict(engineNotes),
            }
          );
          expect(sidebarResp._unsafeUnwrap()).toMatchSnapshot();
          // Expect foo note to have no children
          const fooSidebarItem = sidebarResp._unsafeUnwrap()?.[1];
          expect(fooSidebarItem).toHaveProperty("items", []);
        },
        {
          expect,
          preSetupHook: ENGINE_HOOKS.setupBasic,
        }
      );
    });

    describe("AND nav_exclude_children is set to false", () => {
      test("THEN return sidebar object", async () => {
        await runEngineTestV5(
          async ({ engine }) => {
            const domain = (await engine.getNote("foo")).data!;
            domain.custom.has_collection = true;
            domain.custom.nav_exclude_children = false;
            await engine.writeNote(domain, { metaOnly: true });
            const engineNotes = await engine.findNotes({ excludeStub: false });
            const sidebarResp = getSidebar(
              [{ type: "autogenerated", id: "root" }],
              {
                notes: NoteDictsUtils.createNotePropsByIdDict(engineNotes),
              }
            );
            expect(sidebarResp._unsafeUnwrap()).toMatchSnapshot();
            // Expect foo note to have no children
            const fooSidebarItem = sidebarResp._unsafeUnwrap()?.[1];
            expect(fooSidebarItem).toHaveProperty("items", []);
          },
          {
            expect,
            preSetupHook: ENGINE_HOOKS.setupBasic,
          }
        );
      });
    });
  });

  describe("WHEN nav_exclude is enabled", () => {
    test("THEN return sidebar object", async () => {
      await runEngineTestV5(
        async ({ engine }) => {
          const domain = (await engine.getNote("foo")).data!;
          domain.custom.nav_exclude = true;
          await engine.writeNote(domain, { metaOnly: true });
          const engineNotes = await engine.findNotes({ excludeStub: false });
          const sidebarResp = getSidebar(
            [{ type: "autogenerated", id: "root" }],
            {
              notes: NoteDictsUtils.createNotePropsByIdDict(engineNotes),
            }
          );
          expect(sidebarResp._unsafeUnwrap()).toMatchSnapshot();
          // Expect `foo.md` to be not part of the sidebar
          expect(sidebarResp._unsafeUnwrap()).toHaveLength(1);
        },
        {
          expect,
          preSetupHook: ENGINE_HOOKS.setupBasic,
        }
      );
    });
  });
});
