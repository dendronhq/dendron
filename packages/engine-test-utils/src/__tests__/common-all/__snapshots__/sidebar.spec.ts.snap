// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN sidebar config input WHEN has_collection is enabled AND nav_exclude_children is set to false THEN return sidebar object 1`] = `
Array [
  Object {
    "id": "bar",
    "label": "Bar",
    "type": "note",
  },
  Object {
    "items": Array [],
    "label": "Foo",
    "link": Object {
      "id": "foo",
      "type": "note",
    },
    "type": "category",
  },
]
`;

exports[`GIVEN sidebar config input WHEN has_collection is enabled THEN return sidebar object 1`] = `
Array [
  Object {
    "id": "bar",
    "label": "Bar",
    "type": "note",
  },
  Object {
    "items": Array [],
    "label": "Foo",
    "link": Object {
      "id": "foo",
      "type": "note",
    },
    "type": "category",
  },
]
`;

exports[`GIVEN sidebar config input WHEN nav_exclude is enabled THEN return sidebar object 1`] = `
Array [
  Object {
    "id": "bar",
    "label": "Bar",
    "type": "note",
  },
]
`;

exports[`GIVEN sidebar config input WHEN nav_exclude_children is enabled THEN return sidebar object 1`] = `
Array [
  Object {
    "id": "bar",
    "label": "Bar",
    "type": "note",
  },
  Object {
    "items": Array [],
    "label": "Foo",
    "link": Object {
      "id": "foo",
      "type": "note",
    },
    "type": "category",
  },
]
`;

exports[`GIVEN sidebar config input WHEN providing DefaultSidebar THEN return sidebar object 1`] = `
Array [
  Object {
    "id": "bar",
    "label": "Bar",
    "type": "note",
  },
  Object {
    "items": Array [
      Object {
        "id": "foo.ch1",
        "label": "Ch1",
        "type": "note",
      },
    ],
    "label": "Foo",
    "link": Object {
      "id": "foo",
      "type": "note",
    },
    "type": "category",
  },
]
`;

exports[`GIVEN sidebar config input WHEN providing autogenerated sidebar item THEN return sidebar slice 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "items": Array [
          Object {
            "id": "foo.ch1.gch1.ggch1",
            "label": "Ggch1",
            "type": "note",
          },
        ],
        "label": "Gch1",
        "link": Object {
          "id": "foo.ch1.gch1",
          "type": "note",
        },
        "type": "category",
      },
      Object {
        "id": "foo.ch1.gch2",
        "label": "Gch2",
        "type": "note",
      },
    ],
    "label": "Ch1",
    "link": Object {
      "id": "foo.ch1",
      "type": "note",
    },
    "type": "category",
  },
  Object {
    "id": "foo.ch2",
    "label": "Ch2",
    "type": "note",
  },
]
`;

exports[`GIVEN sidebar config input WHEN providing custom sidebar THEN return sidebar object 1`] = `
Array [
  Object {
    "id": "foo",
    "label": "Foo label",
    "type": "note",
  },
  Object {
    "items": Array [
      Object {
        "items": Array [
          Object {
            "id": "foo.ch1.gch1.ggch1",
            "label": "Ggch1",
            "type": "note",
          },
        ],
        "label": "Gch1",
        "link": Object {
          "id": "foo.ch1.gch1",
          "type": "note",
        },
        "type": "category",
      },
      Object {
        "id": "foo.ch1.gch2",
        "label": "Gch2",
        "type": "note",
      },
      Object {
        "id": "goo.ends-with-ch1.no-ch1-by-itself",
        "label": "Some deep note",
        "type": "note",
      },
    ],
    "label": "Some category label",
    "link": Object {
      "id": "bar",
      "type": "note",
    },
    "type": "category",
  },
]
`;
